### DISCUSSION TITLE
Overture Maps data schema for transportation

### DISCUSSION THREAD

**Discussion by cdbattags** [valhalla/valhalla#4223] [General]
As of yesterday (July 26, 2023), Overture has released an `alpha.0` with a transportation "theme" that uses this [data schema](https://docs.overturemaps.org/themes/transportation/).

I'd like to start the discussion to consider ~~direct support for~~ this format in Valhalla.

EDIT:

Definitely not direct support but maybe a discussion surrounding how everyone is planning to `build_tiles` with this data?

**Comment by cdbattags** [comment]
I've been experimenting today using [osmpbf](https://docs.rs/osmpbf/latest/osmpbf/) to see how easy it might be to reverse engineer how they are ingesting and re-schema-ing the OSM pbfs.

**Comment by kevinkreiser** [comment]
> direct support for this format in Valhalla

i was just bemoaning the inevitable creation of this discussion/issue to my colleagues after i saw the announcement of the release :smile: 

i cannot stress how little enthusiasm i have for such a project. let me explain my position:

1. there is a huge amount of tooling that exists for osm already, it doesnt make sense to add support to each and every one of those tools for this new schema because
2. i, and probably other maintainers, are not really willing to invest the thousands of hours we have sunk into building such aforementioned tooling all over again
3. remember that this is not a new problem, commercial datasets (tomtom, here, zenrin) exist and work with valhalla and other OSM-oriented projects typically via a converter from  whatever external schema to the osm "schema"
4. the effort to write the converter is much more manageable and once its done all the osm tools can use the converted data rather than writing "direct" support in each project

now maybe i'm adding too much emphasis to your use of the word "direct" if that is the case and you are open to using overture converted to osm for use in valhalla, then great we are on the same page. if you really mean "direct" support as in whatever their data looks like we have an executable to parse it into things we can use then lets talk about that a bit more. what does it really mean?

1. write a parser or use a library to access the data in whatever format they have (didnt look at the format yet)
2. figure out a way to iterate over their data in such a way as to fill out our fixed size datastructures that we store in memory mapped files, these are very much modeled after osm
3. port all of the logic from the lua and c++ tag/value parsing to whatever overtures schema is
4. rewrite the whole of the admin parser to use their admins, same problems as above but completely different dataset

this doesnt sound like much but its a huge amount of work. osm, its data model, is deeply ingrained in the existing workings of the algorithms and architecture for the tile building ETL process. trust me that there are many hours of work here. dont get me wrong you are welcome to do it but i have no stomache for it at this point in time.

i think a much more fruitful effort here would be to write a converter from overture to osm pbf.

now maybe a bit of a steelman for your argument, if ive understood it correctly. perhaps there are things in overtures schema that we cannot fully express in osm and as such would be leaving on the table as unsupported features. that would be unfortunate, but before we go through all the work of adding support we had better discuss what those features are and really understand whether or not we can extend osm tagging to support them.

i hope you don't misunderstand my feelings on the subject as antagonistic or a "wet blanket". i really just want to make sure everyone understands what would be involved in directly supporting a new data schema (not just overture but any schema). the long story short is that its a huge amount of work and the same result could likely be achieved by a few folks writing a data converter that everyone's projects can benefit from rather than every project writing their own support for every different input format. i would wager that the companies involved in overture have such a converter already, the question being of course whether or not they will open source it. also, from the press release i read, it seemed to be saying that the transport layer was just ripped out of osm, at least for this version.

**Reply by cdbattags** [reply]
Thank you for your detailed reply! I originally wrote a more thorough post describing my views on the situation and how the political landscape will be hell for a bit with this but decided to go with a more minimal less energized post instead 😅.

I fully agree that direct support is not the answer and that a parser from it into OSM pbf is the best path forward.

I more or less just wanted to start this discussion as it does look like my company will be receiving this format from one of the mapping providers soon and we will be doing the work to shape it for `build_tiles` anyways.

I wouldn't mind my org taking on the first repo for such work and maintenance if this is something the community would want.

EDIT:

> i hope you don't misunderstand my feelings on the subject as antagonistic or a "wet blanket"

Absolutely not received this way!

**Reply by cdbattags** [reply]
> remember that this is not a new problem, commercial datasets (tomtom, here, zenrin) exist and work with valhalla and other OSM-oriented projects typically via a converter from whatever external schema to the osm "schema"

and

> i would wager that the companies involved in overture have such a converter already, the question being of course whether or not they will open source it.

Are the real questions 100%...

**Comment by nilsnolde** [comment]
Jep, 100% agreed with Kevin. Converter is the only way and that's a much broader discussion than Valhalla.

BTW, if you're referring to data from the contact I gave you last week: that's not overture data, that'd come already as OSM PBF.

**Reply by nilsnolde** [reply]
(in case there was a misunderstanding, maybe you're referring to smth else)

**Reply by cdbattags** [reply]
Hahaha I love to see all relevant parties here within an hour of posting this discussion.

> BTW, if you're referring to data from the contact I gave you last week: that's not overture data, that'd come already as OSM PBF.

Thanks for clarifying! Yep, I know that it would be `*.osm.pbf` from you through them:

Map data provider --> [GIS-OPS](https://github.com/gis-ops/prop2osm) --> `*.osm.pbf`

I just wanted to start the larger discussion here for when `<insert-map-data-provider-here>` data is eventually coming to us in Overture format one day 😉 because that's what it seems like, no?

**Reply by nilsnolde** [reply]
>  because that's what it seems like, no?

You can already get the Overture data, that's not the format you'll receive from them. I'll answer on your email, I'm not 100% sure how public that info is already.

**Reply by kevinkreiser** [reply]
> `<insert-map-data-provider-here>` data is eventually coming to us in Overture format one day 😉 because that's what it seems like, no?

id like to chime in on this as well. as a general statement, i definitely think not. the only thing that will be using overture format as their delivery schema will likely be overture itself (meaning the members of their consortium), for the immediate and foreseeable future. i suppose its possible that someone loves it so much that they decide to use the format for their data as well but i wouldnt be willing to claim that it will be the map data standard in any reasonable timespan, i dont care how much big name oomph is behind it.

some food for thought.. did you know that there are already well established existing map data formats that no one outside of the data producers themselves are making use of even though they are backed by the biggest names in the automaker industry? you can google `NDS`, maybe learn a bit about its features, take note that its light years ahead of overture and then ask yourself why you've never heard of it :smile:

anyway, just lookign at the stuff @nilsnolde linked i get strong vibes of how incumbent commercial datasets (eg HERE) are already delivered with the way attribution is linearly referenced along each segment. in my opinion, the novel thing about overture is the opening up of a ton of new data/coverage. stuff that would be hard to add to osm without angering the osm editors themselves (people dont like mass imports especially if you put big corporate names on them haha).

anyway again just my 2 cents!

**Reply by cdbattags** [reply]
> some food for thought.. did you know that there are already well established existing map data formats that no one outside of the data producers themselves are making use of even though they are backed by the biggest names in the automaker industry? you can google NDS, maybe learn a bit about its features, take note that its light years ahead of overture and then ask yourself why you've never heard of it 😄

Fully agree and that's why we'll continue to use OSM as a base layer always with the ID references. I think the intriguing part here is that maybe the Overture foundation members will somewhat try to align their data with each other in a way that makes combining them not a total PITA. Obviously, licenses will need to be considered but I'm hoping that we might one day reach a spot where the commercial players align on the same IDs for their nodes/ways.

**Reply by cdbattags** [reply]
In reference to our work at https://tripleaxle.com we have a massive amount of state-specific restriction data that we are going to layer on top of OSM and then looking to purchase one of the commercial players as well for the general non-permit truck restrictions.

All the data we find we play to align it directly to OSM first and then use the commercial guys "on top" but the bones are always going to be OSM.

**Comment by gknisely** [comment]
I was waiting for this one to pop up.  I agree a converter is the best option here.

**Comment by nilsnolde** [comment]
FWIW, I tried yesterday quickly already.. I had a deeper look into their schema ~~and so far it's hardly usable for proper routing~~ (that's not true, discovered a few things later): https://docs.overturemaps.org/reference/transportation/segment. It's a simple-feature model, no relations, which is nice. And admittedly they're further than I figured. Seems quite useless for proper guidance/navigation, but proper routing seems quite possible. One thing I couldn't see was multi-way turn restrictions. Converting that to OSM PBF & schema will be quite a PITA though :smile:

**Reply by kevinkreiser** [reply]
they have a turn restriction example and in the json you can see the `via` part is an array, so i am pretty sure you just put multiple vias in that array to do multiple way restriction
